src/MySqlConnector/Serialization/MySqlSession.cs

```diff
@@ -526,47 +526,81 @@ private void VerifyConnected()
-					throw new MySqlException("Either the SSL Certificate Password is incorrect or the SSL Certificate File is invalid", ex);
+						throw new MySqlException("Cannot find Certificate File", ex);
+					throw new MySqlException("Either the Certificate Password is incorrect or the Certificate File is invalid", ex);
+				}
+			}
+
+			X509Chain caCertificateChain = null;
+			if (cs.CACertificateFile != null)
+			{
+				try
+				{
+					var caCertificate = new X509Certificate2(cs.CACertificateFile);
+					caCertificateChain = new X509Chain
+					{
+						ChainPolicy =
+						{
+							RevocationMode = X509RevocationMode.NoCheck,
+							VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority
+						}
+					};
+					caCertificateChain.ChainPolicy.ExtraStore.Add(caCertificate);
+				}
+				catch (CryptographicException ex)
+				{
+					if (!File.Exists(cs.CACertificateFile))
+						throw new MySqlException("Cannot find CA Certificate File", ex);
+					throw new MySqlException("The CA Certificate File is invalid", ex);
 				}
 			}
 
-			Func<object, string, X509CertificateCollection, X509Certificate, string[], X509Certificate> localCertificateCb =
-				(lcbSender, lcbTargetHost, lcbLocalCertificates, lcbRemoteCertificate, lcbAcceptableIssuers) => lcbLocalCertificates[0];
+			X509Certificate LocalCertificateCb(object lcbSender, string lcbTargetHost, X509CertificateCollection lcbLocalCertificates, X509Certificate lcbRemoteCertificate, string[] lcbAcceptableIssuers) => lcbLocalCertificates[0];
+
+			bool RemoteCertificateCb(object rcbSender, X509Certificate rcbCertificate, X509Chain rcbChain, SslPolicyErrors rcbPolicyErrors)
+			{
+				if (cs.SslMode == MySqlSslMode.Preferred || cs.SslMode == MySqlSslMode.Required)
+					return true;
 
-			Func<object, X509Certificate, X509Chain, SslPolicyErrors, bool> remoteCertificateCb =
-				(rcbSender, rcbCertificate, rcbChain, rcbPolicyErrors) =>
+				if ((rcbPolicyErrors & SslPolicyErrors.RemoteCertificateChainErrors) != 0 && caCertificateChain != null)
 				{
-					switch (rcbPolicyErrors)
+					if (caCertificateChain.Build((X509Certificate2) rcbCertificate))
 					{
-						case SslPolicyErrors.None:
-							return true;
-						case SslPolicyErrors.RemoteCertificateNameMismatch:
-							return cs.SslMode != MySqlSslMode.VerifyFull;
-						default:
-							return cs.SslMode == MySqlSslMode.Preferred || cs.SslMode == MySqlSslMode.Required;
+						var chainStatus = caCertificateChain.ChainStatus[0].Status & ~X509ChainStatusFlags.UntrustedRoot;
+						if (chainStatus == X509ChainStatusFlags.NoError)
+							rcbPolicyErrors &= ~SslPolicyErrors.RemoteCertificateChainErrors;
 					}
-				};
+				}
+
+				if (cs.SslMode == MySqlSslMode.VerifyCA)
+					rcbPolicyErrors &= ~SslPolicyErrors.RemoteCertificateNameMismatch;
+
+				return rcbPolicyErrors == SslPolicyErrors.None;
+			}
 
 			SslStream sslStream;
-			if (clientCertificates == null)
+			if (CertificateFiles == null || CertificateFiles.Count == 0)
```

Comment:
Right, shouldn't happen. It just doesn't feel right referring to the element [0] without checking the count, but it has worked fine without this check so far. I'll take it out.